tamr_client/datasets/record.py

```diff
@@ -0,0 +1,124 @@
+"""
+See https://docs.tamr.com/reference/record
+"The recommended approach for interacting with records is to use the upsert and delete functions for all use cases they
+can handle. For more advanced use cases, the underlying _update function can be used directly."
+"""
+import json
+from typing import Dict, Iterable, Union
+
+import tamr_client as tc
+from tamr_client.types import JsonDict
+
+
+class PrimaryKeyNotFound(Exception):
+    """Raised when referencing a primary key by name that does not exist."""
+
+    pass
+
+
+def _update(
+    session: tc.Session, dataset: tc.Dataset, updates: Iterable[Dict]
+) -> JsonDict:
+    """Send a batch of record creations/updates/deletions to this dataset.
+    You probably want to use :func:`~tamr_client.dataset.upsert_records`
+    or :func:`~tamr_client.dataset.delete_records` instead.
+
+    Args:
+        dataset: Dataset containing records to be updated
+        updates: Each update should be formatted as specified in the `Public Docs for Dataset updates <https://docs.tamr.com/reference#modify-a-datasets-records>`_.
+
+    Returns:
+        JSON response body from server
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered
+    """
+    stringified_updates = (json.dumps(update) for update in updates)
+    r = session.post(
+        str(dataset.url) + ":updateRecords",
+        headers={"Content-Encoding": "utf-8"},
+        data=stringified_updates,
+    )
+    return tc.response.successful(r).json()
+
+
+def upsert(
+    session: tc.Session,
+    dataset: tc.Dataset,
+    records: Iterable[Dict],
+    *,
+    primary_key_name: str,
```

Comment:
Why does `primary_key_name` need to be specified? This can be determined by interrogating the dataset.

Reply:
The motivation here is to ensure that the primary key name in the dataset and in the data match.  If this is not true, the user needs to coerce their data column names to match Tamr.  I was aiming to prioritize avoiding unintentional actions over convenience, since I got the impression from the field that this would be preferred.

Reply:
Yea I think the argument is that the data schema *is* part of the interface and doing a column rename to align the primary keys is easy for a user to do.